x/stakeibc/keeper/icacallbacks_delegate_test.go

```diff
@@ -0,0 +1,270 @@
+	s.Require().NoError(err)
+
+	return DelegateCallbackTestCase{
+		initialState: DelegateCallbackState{
+			stakedBal:		stakedBal,
+			balanceToStake: balanceToStake,
+			depositRecord:	depositRecord,
+			callbackArgs:	callbackArgs,
+			val1Bal: 		val1Bal,
+			val2Bal: 		val2Bal,
+			val1RelAmt:		val1RelAmt,
+			val2RelAmt:		val2RelAmt,
+		},
+		validArgs: DelegateCallbackArgs{
+			packet:   packet,
+			ack: 	  &ack,
+			args:     args,
+		},
+	}
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_Successful() {
+	tc := s.SetupDelegateCallback()
+	initialState := tc.initialState
+	validArgs := tc.validArgs
+
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
+	s.Require().NoError(err)
+
+	// Confirm stakedBal has increased
+	hostZone, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx(), chainId)
+	s.Require().True(found)
+	s.Require().Equal(int64(initialState.stakedBal) + initialState.balanceToStake, int64(hostZone.StakedBal), "stakedBal should have increased")
+
+	// Confirm delegations have been added to validators
+	val1 := hostZone.Validators[0]
+	val2 := hostZone.Validators[1]
+	s.Require().Equal(int64(initialState.val1Bal) + initialState.val1RelAmt, int64(val1.DelegationAmt), "val1 balance should have increased")
+	s.Require().Equal(int64(initialState.val2Bal) + initialState.val2RelAmt, int64(val2.DelegationAmt), "val2 balance should have increased")
+	
+	// Confirm deposit record has been removed
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 0, "number of deposit records")
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_DelegateCallbackTimeout() {
+	tc := s.SetupDelegateCallback()
+	validArgs := tc.validArgs
+	// a nil ack means the request timed out
+	validArgs.ack = nil
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
+	s.Require().NoError(err)
+
+	// Confirm stakedBal has not increased
+	hostZone, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx(), chainId)
+	s.Require().True(found)
+	s.Require().Equal(int64(tc.initialState.stakedBal), int64(hostZone.StakedBal), "stakedBal should not have increased")
+
+	// Confirm deposit record has been removed
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 1, "number of deposit records")
+	record := records[0]
+	s.Require().Equal(recordtypes.DepositRecord_STAKE, record.Status, "deposit record status should be timeout")
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_DelegateCallbackErrorOnHost() {
+	tc := s.SetupDelegateCallback()
+	validArgs := tc.validArgs
+	// an error ack means the tx failed on the host
+	fullAck := channeltypes.Acknowledgement{Response: &channeltypes.Acknowledgement_Error{Error: "error"}}
+	validArgs.ack = &fullAck
+
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
```

Comment:
shouldn't this return an error?

Reply:
the code does return nil for this case https://github.com/Stride-Labs/stride/blob/d66d3d4f026ba65c93047b00476769d4173b4a2d/x/stakeibc/keeper/icacallbacks_delegate.go#L54

from a design pov, this is the case where a tx has failed on the host chain - I think this is a valid state transition on stride (we should be able to handle errors on the host chain and update state on stride accordingly)

wdyt?

Reply:
Ah nvm, I thought this was covering this case:
```
	txMsgData, err := icacallbacks.GetTxMsgData(ctx, *ack, k.Logger(ctx))
	if err != nil {
		k.Logger(ctx).Error(fmt.Sprintf("failed to fetch txMsgData, packet %v", packet))
		return sdkerrors.Wrap(icacallbackstypes.ErrTxMsgData, err.Error())
	}
```
Agreed nil should be returned!